HPCC-21524 Check thor log restarting when retrieving WU thor log

wangkx opened the pull request:
If thor master/slave is restarted before a WU is finished, the
"Finished wuid=..." will not be in thor log. The existing code
for retrieving WU thor log only uses the "Finished wuid=..." to
find out the end of the WU thor log. In this fix, we also check
the thor log restarting (by checking log ID) for the end of the
WU thor log.

Also adding code to retrieve WU thor logs using a thread pool
for a better performance.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-21524 Check thor log restarting when retrieving WU thor log

If thor master/slave is restarted before a WU is finished, the
"Finished wuid=..." will not be in thor log. The existing code
for retrieving WU thor log only uses the "Finished wuid=..." to
find out the end of the WU thor log. In this fix, we also check
the thor log restarting (by checking log ID) for the end of the
WU thor log.

Also adding code to retrieve WU thor logs using a thread pool
for a better performance.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@afishbeck please review.

HPCCSmoketest commented:
This PR (12217) is in the 6/6 position of the queue. Estimated start time is after ~3.0 hour(s)

HPCCSmoketest commented:
This PR (12217) is in the 5/5 position of the queue. Estimated start time is after ~2.5 hour(s)

HPCCSmoketest commented:
This PR (12217) is in the 4/4 position of the queue. Estimated start time is after ~2.0 hour(s)

HPCCSmoketest commented:
This PR (12217) is in the 3/3 position of the queue. Estimated start time is after ~1.5 hour(s)

HPCCSmoketest commented:
This PR (12217) is in the 2/2 position of the queue. Estimated start time is after ~1.0 hour(s)

HPCCSmoketest commented:
This PR (12217) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12217, label: h21524 starts now.
The reason of this test is: Did not build yet.
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6e9ec61b61acd2f0c8811cd394d98b6d20e4d905
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |815 |815 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |889 |889 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   41 sec (00:00:41) |  184 sec (00:03:04) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1432 sec (00:23:52) |   19 sec (00:00:19) | 1696 sec (00:28:16) |

HPCCSmoketest commented:
Process of PR-12217, label: h21524 starts now.
The reason of this test is: Base branch changed.
Commit ID: 6e9ec61b61acd2f0c8811cd394d98b6d20e4d905
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6e9ec61b61acd2f0c8811cd394d98b6d20e4d905
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |815 |815 |0 |247 sec  (00:04:07) |
|test (thor) |738 |738 |0 |579 sec  (00:09:39) |
|test (roxie) |889 |889 |0 |344 sec  (00:05:44) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  181 sec (00:03:01) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1426 sec (00:23:46) |   21 sec (00:00:21) | 1678 sec (00:27:58) |

afishbeck commented:
@jakesmith as this relates to Thor log files, do you want to take a look?

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2092,52 +2055,75 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
 
-        VStringBuffer startwuid("Started wuid=%s", wuid.str());
-        VStringBuffer endwuid("Finished wuid=%s", wuid.str());
+void WsWuInfo::readWorkunitLog(OwnedIFileIOStream ios, MemoryBuffer& buf, const char* outFile)
```
this is going 'copy' the parameter passed. In this case, it will link the parameter passed,
then when 'ios' goes out of scope it will release it.

Did you intend it to be like this?
I think there are -ve leaks where it is called.

Normally you'd expect the method to take a regular unowned object, i.e. IFileIOStream * in this case.
and for the ownership to be managed elsewhere.

wangkx replied:
Good catch. I will fix it.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1985,47 +1987,8 @@ void WsWuInfo::getWorkunitThorLog(const char* fileName, MemoryBuffer& buf, const
-        if (strstr(line.str(), ew))
-            include = false;
-    }
+    readWorkunitLog(createBufferedIOStream(rIO), buf, outFile);
```
This is going to leak...
Because readWorkunitLog(OwnedIFileIOStream ios, is going to link/release itself.
Nothing is owns the object being created here, so it will be leaked.

See other comments re. readWorkunitLog(OwnedIFileIOStream, issue.

wangkx replied:
Will fix

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2092,52 +2055,75 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
-            line.clear();
-            for (;;)
+            char c;
+            size32_t numRead = ios->read(1, &c);
```
this is not going to be efficient.
If there isn't already, should add a utility function to jlib somewhere that reads a line (efficiently) from a stream.
e..g reads a buffer at a time.

wangkx replied:
ok

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2092,52 +2055,75 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
-        const char *sw = startwuid.str();
-        const char *ew = endwuid.str();
+    const char *sw = startwuid.str();
+    const char *ew = endwuid.str();
```
trivial: don't need .str(); cast operator will.
Not sure it's worth assigning to sw/eq either, it's only 1 virtual call per line (the ios->read per character is far more of an issue).

wangkx replied:
Will fix.

jakesmith commented:
@wangkx - back to you.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -3515,12 +3506,18 @@ void CWsWuFileHelper::createThorSlaveLogfile(Owned<IConstWorkUnit>& cwu, WsWuInf
         }
     }
+    ForEachItemIn(i, threadHandles)
+        threadPool->join(threadHandles.item(i));
```
Will change it to joinAll.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -641,5 +643,83 @@ class CWsWuEmailHelper
+    }
+};
+
+class CGetThorSlaveLogToFileThread : public CInterface, implements IPooledThread
```
Will change it to CSimpleInterfaceOf.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -641,5 +643,83 @@ class CWsWuEmailHelper
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CGetThorSlaveLogToFileThread() {};
```
Will remove.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -641,5 +643,83 @@ class CWsWuEmailHelper
+    IMPLEMENT_IINTERFACE;
+
+    CGetThorSlaveLogToFileThreadParam(WsWuInfo* _wuInfo, CThorSlaveLogFileItem* _logFile)
+        : wuInfo(_wuInfo)
```
Will fix.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -3515,12 +3506,18 @@ void CWsWuFileHelper::createThorSlaveLogfile(Owned<IConstWorkUnit>& cwu, WsWuInf
+                    logDate, logDir, i+1, fileName);
+                Owned<CGetThorSlaveLogToFileThreadParam> threadParam = new CGetThorSlaveLogToFileThreadParam(
+                    &winfo, logFile.getClear());
+                PooledThreadHandle handle = threadPool->start( threadParam.getClear());
```
I will remove the space.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -39,6 +39,8 @@
 #include "zcrypt.hpp"
 #endif
 
+static const int THREAD_POOL_SIZE = 40;
```
I will make it to be configurable.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -3515,12 +3506,18 @@ void CWsWuFileHelper::createThorSlaveLogfile(Owned<IConstWorkUnit>& cwu, WsWuInf
                 VStringBuffer fileName("%s%c%s_thorslave.%u.%s.log", path, PATHSEPCHAR, processName.str(), i+1, logDate.str());
-                winfo.getWorkunitThorSlaveLog(groupName.str(), NULL, logDate.str(), logDir.str(), i+1, mb, fileName.str(), false);
+                Owned<CThorSlaveLogFileItem> logFile = new CThorSlaveLogFileItem(groupName,
+                    logDate, logDir, i+1, fileName);
```
Will pass all params.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2092,52 +2055,75 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
+            //So, we should check whether the slave is restarting or not.
+            //If restarting, the row ID will be reset to a small number (assume < 5: I saw 0, 1, and 2).
+            unsigned long rowID;
+            if (readLogLineID((char*)line.str(), rowID) && (rowID < 5))
```
A JIRA has been created to make sure that WU ZAP report contains the log files if WU is still running at midnight and continues in a new log.

I thought about using process ID. It takes more time to parse. But, may be worth to do.

jakesmith replied:
>A JIRA has been created to make sure that WU ZAP report contains the log files if WU is still running at midnight and continues in a new log.

Can you add the JIRA URL here for reference? (and link it as related in this PR's JIRA too).

wangkx replied:
https://track.hpccsystems.com/browse/HPCC-21753

wangkx committed:
HPCC-21524 Revise based on review

Call joinAll();
Use CSimpleInterfaceOf where possible;
Make the thread pool size to be configurable and default to 100;
Use process ID to check server restart;
Fix a possible memory leak;
Add readLine function to jlib and call it to read log files;
Other minor changes.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review my changes.

HPCCSmoketest commented:
Process of PR-12217, label: h21524 starts now.
The reason of this test is: New commit.
Commit ID: 04dce846ebbe70833cbd243760ddd198079ebadb
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 04dce846ebbe70833cbd243760ddd198079ebadb
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |31 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |815 |815 |0 |179 sec  (00:02:59) |
|test (thor) |738 |738 |0 |629 sec  (00:10:29) |
|test (roxie) |889 |889 |0 |217 sec  (00:03:37) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1305 sec (00:21:45) |   21 sec (00:00:21) | 1554 sec (00:25:54) |

jakesmith commented on system/jlib/jio.ipp:
```diff
@@ -99,6 +99,80 @@ protected:
+            size32_t tempOffset = curBufferOffset;
+            while (tempOffset < numInBuffer)
+            {
+                if (buffer[tempOffset] == '\r') // standard case, \r\n marks a new header line.
```
minor: would probably be a bit more efficient to use a pointer to the buffer that you increment here.

wangkx replied:
Will change to a pointer.

jakesmith commented on system/jlib/jio.ipp:
```diff
@@ -99,6 +99,80 @@ protected:
+                        }
+
+                        outBuffer.append(outLength, outStartFrom);
+                        curBufferOffset = tempOffset;
```
if you used a pointer, you could calc. length by doing ptr-start and avoid outLength and tempOffset

jakesmith commented on system/jlib/jio.ipp:
```diff
@@ -99,6 +99,80 @@ protected:
+            }
+
+            if(foundCRLF)
+                break;
```
There's a bit of inconsistency here, of when eof is returned from readLineFromBuffer.
i.e. only checked if needs to read 1 more char following \r.

It should be consistent.

wangkx replied:
Do you want me to check a possible eof after the CRLF is found (even if it is not the case of reading 1 more char following \r)?

jakesmith commented:
I don't think the approach taken (in the latest commit) that implements the readLine is quite right..
i.e. the new jlib IBufferedFileIOStream interface
and IBufferedFileIOStream * createBufferedFileIOStream(IFileIO * file,...

That's quite confusing, we already had a factory method to create a buffered implementation from a IFileIO, i.e. "IFileIOStream * createBufferedIOStream(IFileIO * file, ..."

The new interface isn't so much a buffered io, as a supplemented IFileIOStream with read line capabilities.

It would be better to avoid extending this specific (file) interface and have a generalized mechanism that worked on any stream type (e.g. IIOStream or ISimpleReadStream)

wangkx committed:
HPCC-21524 WIP: revise based on review

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-12217, label: h21524 starts now.
The reason of this test is: New commit.
Commit ID: 47adb206b18166291570c324205f61b75d0cc4d4
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 47adb206b18166291570c324205f61b75d0cc4d4
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |32 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |815 |815 |0 |158 sec  (00:02:38) |
|test (thor) |738 |738 |0 |640 sec  (00:10:40) |
|test (roxie) |889 |889 |0 |198 sec  (00:03:18) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  186 sec (00:03:06) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1303 sec (00:21:43) |   19 sec (00:00:19) | 1558 sec (00:25:58) |

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1859,19 +1859,9 @@ void appendIOStreamContent(MemoryBuffer &mb, IFileIOStream *ios, bool forDownloa
-                break;
-        }
+        eof = ios->readLine(true, line);
+        if (line.isEmpty())
```
I'm not sure this should be possible.
If empty, eof should be set I think.

wangkx replied:
Yes with the IStreamLineReader.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1985,47 +1965,9 @@ void WsWuInfo::getWorkunitThorLog(const char* fileName, MemoryBuffer& buf, const
-            include = false;
-    }
+    Owned<IBufferedFileIOStream> ios = createBufferedFileIOStream(rIO);
+    readWorkunitLog(ios.get(), buf, outFile);
```
don't need the .get() - operator-> will deference.

wangkx replied:
ok

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2089,53 +2031,71 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
+    }
+}
+
+void WsWuInfo::readWorkunitLog(IBufferedFileIOStream* ios, MemoryBuffer& buf, const char* outFile)
```
is 'buf' ever used?

wangkx replied:
Not for downloading ZAP report. But, it may be used for viewing a file content.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2089,53 +2031,71 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
-            CWsWuFileHelper helper(nullptr);
-            outIOS.setown(helper.createIOStreamWithFileName(outFile, IFOcreate));
+            outputThisLine = true;
+            outputALine(line.length(), line.str(), buf, outIOS);
```
see question about 'buf' above. Afaics, readWorkunitLog is only ever called with a stream, buf is not used.

If so, it would clearer to call:
```
        outIOS->write(line.length(), line.str());
```
directly here

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2089,53 +2031,71 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
+
+    char* eptr = nullptr;
+    const char* ids = items.item(3);
+    unsigned processID = (unsigned)strtol(ids, &eptr, 10);
```
Use of atoi() is more standard.

jakesmith replied:
Also, this is also quite expensive. ( Copying everyline into N strings, and putting them in an array ).

It would be better to scan to field, and to end of field, then use atoi_l()

wangkx replied:
I am using strtol because it can check any character inside the field just in case.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -3703,12 +3669,14 @@ int CWsWuFileHelper::zipAFolder(const char* folder, bool gzip, const char* zipFi
 
 void CWsWuFileHelper::createWUZAPFile(IEspContext& context, Owned<IConstWorkUnit>& cwu, CWsWuZAPInfoReq& request,
-    StringBuffer& zipFileName, StringBuffer& zipFileNameWithPath)
+    StringBuffer& zipFileName, StringBuffer& zipFileNameWithPath, unsigned _getThorSlaveLogThreadPoolSize)
```
> _getThorSlaveLogThreadPoolSize

'get' in variable name is misleading.

remove from here and other similar places.

wangkx replied:
ok

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -3703,12 +3669,14 @@ int CWsWuFileHelper::zipAFolder(const char* folder, bool gzip, const char* zipFi
 }
 
 void CWsWuFileHelper::createWUZAPFile(IEspContext& context, Owned<IConstWorkUnit>& cwu, CWsWuZAPInfoReq& request,
```
>Owned<IConstWorkUnit>& cwu

Not new, but shouldn't be passing Owned<...> references around in general.
Method should take underlying type (e.g. IConstWorkUnit *). You can pass in the Owned<...> object and it will automatically be deferenced to the underlying type.

wangkx replied:
Will switch to a pointer.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -139,6 +141,8 @@ class WsWuInfo
     void readArchiveFiles(IPropertyTree* archiveTree, const char* path, IArrayOf<IEspWUArchiveFile>& files);
     void outputALine(size32_t len, const char* content, MemoryBuffer& outputBuf, IFileIOStream* outIOS);
+    unsigned readLogProcessID(const char* linePtr);
+    void readWorkunitLog(IBufferedFileIOStream* ios, MemoryBuffer& buf, const char* outFile);
```
see other comment, not sure 'MemoryBuffer& buf' is used, should be removed if not.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -617,8 +622,9 @@ class CWsWuFileHelper
-        StringBuffer &zipFileName, StringBuffer &zipFileNameWithPath);
-    IFileIOStream* createWUZAPFileIOStream(IEspContext &context, Owned<IConstWorkUnit> &cwu, CWsWuZAPInfoReq &request);
+        StringBuffer &zipFileName, StringBuffer &zipFileNameWithPath, unsigned _getThorSlaveLogThreadPoolSize);
+    IFileIOStream* createWUZAPFileIOStream(IEspContext &context, Owned<IConstWorkUnit> &cwu,
```
not new: see other comment re. not using Owned<..> &obj parameters.

wangkx replied:
Will switch to a pointer.

jakesmith commented on esp/services/ws_workunits/ws_workunitsService.hpp:
```diff
@@ -361,6 +362,12 @@ class CWsWorkunitsSoapBindingEx : public CWsWorkunitsSoapBinding
         batchWatchFeaturesOnly = cfg->getPropBool(xpath.str(), false);
         directories.set(cfg->queryPropTree("Software/Directories"));
+
+        xpath.setf("Software/EspProcess[@name=\"%s\"]/EspBinding[@name=\"%s\"]/service", process, name);
```
not for now: really anything like this that builds up an xpath to probe the environment or similar, should move to somewhere common, e.g. into environment.*, e.g. IPropertyTree *getEnvServiceBinding(process, serviceName)

wangkx replied:
I need both service and binding here. I am going to create:
IPropertyTree *getEnvServiceBinding(process, bindingName);
IPropertyTree *getEnvServiceBinding(process, serviceName);

jakesmith replied:
>the xpath here is for reading esp.xml

ok, I didn't realize that, it looks like a clone of the structures from environment.xml sections.

Either way, I would look at this after this separately, after this PR is complete.
There are probably other places that would similarly benefit from moving the manually created xpaths to helper methods in environment.*

@wangkx

jakesmith commented on initfiles/componentfiles/configxml/@temp/esp_service_WsSMC.xsl:
```diff
@@ -230,6 +230,9 @@ This is required by its binding with ESP service '<xsl:value-of select="$espServ
                 <ClusterQueryStateThreadPoolSize><xsl:value-of select="@clusterQueryStateThreadPoolSize"/></ClusterQueryStateThreadPoolSize>
             </xsl:if>
+            <xsl:if test="string(@GetThorSlaveLogThreadPoolSize) != ''">
+                <GetThorSlaveLogThreadPoolSize><xsl:value-of select="@GetThorSlaveLogThreadPoolSize"/></GetThorSlaveLogThreadPoolSize>
```
'Get' isn't appropriate here and other places.

jakesmith commented on system/jlib/jutil.hpp:
```diff
@@ -231,9 +231,9 @@ class jlib_decl StringArray : public ArrayOf<const char *, const char *, StringP
+    void appendList(const char *list, const char *delim, bool trimSpaces = true, unsigned maxNumItems = 0);//maxNumItems = 0: no limit
     // Appends a list in a string delimited by 'delim' without duplicates
-    void appendListUniq(const char *list, const char *delim, bool trimSpaces = true);
+    void appendListUniq(const char *list, const char *delim, bool trimSpaces = true, unsigned maxNumItems = 0);
```
not convinced about this change (and changed to DelimToStringArray).

See other comments about parsing out the processID more efficiently. In which case use of this will not be necessary.

wangkx commented:
I will create another PR for 7.2.x.
